sycl/include/CL/sycl/half_type.hpp

```diff
@@ -86,22 +88,132 @@ inline float cast_if_host_half(half_impl::half val) {
 } // namespace sycl
 } // namespace cl
 
+#ifdef __SYCL_DEVICE_ONLY__
+using half = _Float16;
+#else
+using half = cl::sycl::detail::half_impl::half;
+#endif
+
+// Partial specialization of some functions in namespace `std`
 namespace std {
 
-template <> struct hash<cl::sycl::detail::half_impl::half> {
-  size_t operator()(cl::sycl::detail::half_impl::half const &key) const
-      noexcept {
-    return hash<uint16_t>()(key.Buf);
+#ifdef __SYCL_DEVICE_ONLY__
+// `constexpr` could work because the implicit conversion from `float` to
+// `_Float16` can be `constexpr`.
+#define CONSTEXPR_QUALIFIER constexpr
+#else
+// The qualifier is `const` instead of `constexpr` that is original to be
+// because the constructor is not `constexpr` function.
+#define CONSTEXPR_QUALIFIER const
+#endif
+
+// Partial specialization of `std::hash<cl::sycl::half>`
+template <> struct hash<half> {
+  size_t operator()(half const &Key) const noexcept {
+    return hash<uint16_t>()(reinterpret_cast<const uint16_t &>(Key));
```

Comment:
Curious `()`...
Does this compile?

Reply:
Ah I see, I was confused by the old C++ style.
What about
```
     return hash<uint16_t> {} (reinterpret_cast<const uint16_t &>(Key));
```
to make clearer (well, to me...) that you have a functor constructor :-)

Reply:
Are you testing for this in the tests somewhere?

Reply:
You’re right. There is no case for that before. I’ll add it.